### PR TITLE
Arteria runfolder handlers

### PR DESCRIPTION
**Pull request by nkongenelly**

### CODE REVIEW

**Review by Aratz — COMMENTED** [review]
Great work! I left a few comments and suggestions for you to address. I think the main thing is that we should really try to decouple `Runfolder` from the service's interface, i.e. there shouldn't be anything web specific in that class.

File: arteria/models/runfolder_utils.py [hunk 1]
```diff
@@ -76,8 +92,9 @@ def state(self):
 
     @state.setter
     def state(self, new_state):
```
[the remark this answers is not in the document]
Comment (Aratz) [reply arteria/models/runfolder_utils.py:94]: Actually, here maybe here it would be nice to use type annotation to make this clearer

**Comment by nkongenelly** [comment]
@Aratz Thank you so much for the review. I have learnt a lot while making the corrections.

I have now updated the code:

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    runfolder_dict['host'], runfolder_dict['link'] = get_host_link(request)
+
+    return runfolder_dict
+
+
+def set_exceptions(e, kwargs=None):
+    if type(e) is AssertionError:
+        raise web.HTTPNotFound(reason=e.args[0] if len(e.args) > 0 else e)
+    else:
+        raise e
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:156]: If the goal is to have more information available for debugging, I think the code would be clearer to not have such a function and instead have the following code for each function. I think it's better if one sees which exception can be raised by the handler at once. (you'll need import `logging` and add `log = logging.getLogger(__name__)` at the begging of this file).

```python
except AssertionError as exc:
    log.exception("")
    raise web.HTTPNotFound from exc
```

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    """
+    Returns some information about the runfolder as json
+    """
+    try:
+        runfolder_name = request.match_info['runfolder']
+        runfolders = [
+
+            serialize_runfolder_path(Runfolder(Path(runfolder_path) / runfolder_name), request)
+            for runfolder_path in Config().get('monitored_directories')
+        ]
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:45]: This will not work, I've updated the tests to emphasize what the problem is.

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+
+@routes.get("/runfolders/next")
+async def get_next_runfolder(request):
+    """
+    Finds unprocessed runfolder (state=ready) and then
+    returns some information about this runfolder.
+    """
+    try:
+        runfolder_cls = list_runfolders(
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:62]: This variable should be named something else, to emphasize that it's a list of runfolders. For instance `runfolders`.

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            Config()['monitored_directories'],
+            filter_key=lambda r: r.state == State.READY
+        )
+
+        if len(runfolder_cls) > 0:
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:67]:
```suggestion
        if runfolders_cls:

```

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            runfolder_dict0 = serialize_runfolder_path(runfolder_cls[0], request)
+
+            return web.json_response(
+                data=runfolder_dict0,
+                status=200
+            )
+        else:
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:74]: You can add a log message here to mention no runfolders were found.

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        set_exceptions(e)
+
+
+@routes.get("/runfolders/pickup")
+async def get_pickup_runfolder(request):
+    """
+    Used to start processing runfolders and also sets the runfolder to PENDING state.
+    """
+    try:
+        runfolder_cls = list_runfolders(
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:88]: Same comments as `get_next_runfolder`

File: arteria/handlers/arteria_runfolder_handlers.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+
+def serialize_runfolder_path(runfolder_cls, request):
+    """
+    Get the path uri as web.json_response gives an error when
+    self.path is of type Path
+    """
+    runfolder_dict = runfolder_cls.to_dict()
+    runfolder_dict['path'] = Path(runfolder_dict['path']).as_uri()
+    runfolder_dict['host'], runfolder_dict['link'] = get_host_link(request)
```
Comment (Aratz) [inline arteria/handlers/arteria_runfolder_handlers.py:147]: I missed that the first time, but `link` should be a link to the `path` endpoint in case one wants more info about this runfolder, e.g. ` http://mm-xart002:10800/api/1.0/runfolders/path/data/mm-xart002/runfolders/180718_M04499_0151_000000000-BWFHW` (for all endpoints that return a runfolder).

**Review by Aratz — COMMENTED** [review]
Nice :) There are still some issues that need fixing. I've updated the tests and will push shortly. There are also a few comments from the previous review that were not addressed. Please do so or otherwise give a reason why not.

File: arteria/models/runfolder_utils.py [hunk 9]
```diff
@@ -34,14 +35,19 @@ def list_runfolders(monitored_directories, filter_key=lambda r: True):
...
     A class to manipulate runfolders on disk
     """
     def __init__(self, path):
         self.config = Config(DEFAULT_CONFIG)
         self.path = Path(path)
-        assert self.path.is_dir()
+
+        runfolder_name = os.path.basename(self.path)
+        if not self.path.is_dir():
+            raise AssertionError(f"Runfolder '{runfolder_name}' does not exist")
```
Comment (Aratz) [inline arteria/models/runfolder_utils.py:49]:
```suggestion
        assert self.path.is_dir(), f"Runfolder '{runfolder_name}' does not exist"
```

File: arteria/models/runfolder_utils.py [hunk 10]
```diff
@@ -34,14 +35,19 @@ def list_runfolders(monitored_directories, filter_key=lambda r: True):
...
 class Runfolder():
     """
     A class to manipulate runfolders on disk
     """
     def __init__(self, path):
         self.config = Config(DEFAULT_CONFIG)
         self.path = Path(path)
-        assert self.path.is_dir()
+
+        runfolder_name = os.path.basename(self.path)
```
Comment (Aratz) [inline arteria/models/runfolder_utils.py:47]:
```suggestion
        runfolder_name = self.path.name
```

File: arteria/models/runfolder_utils.py [hunk 11]
```diff
@@ -34,14 +35,19 @@ def list_runfolders(monitored_directories, filter_key=lambda r: True):
```
Comment (Aratz) [inline arteria/models/runfolder_utils.py:35]: I don't think the if statement is necessary in the `except-`block above, i.e. if the `if-`statement is false, is will also continue to the next iteration of the loop.

If this is the behavior you want, the code should probably be:
```python
            except AssertionError as e:
                continue
```

If you want to forward the exception if it contains a different message, the code should probably look something like below, and you should add a unit test to check the exception is forwarded correctly
```python
            except AssertionError as e:
                if str(e) != f"File [Rr]unParameters.xml not found in runfolder {subdir}":
                    raise
```

File: arteria/models/runfolder_utils.py [hunk 12]
```diff
@@ -116,6 +125,17 @@ def metadata(self):
 
         return metadata
 
+    def __setitem__(self, key, value):
+        return setattr(self, key, value)
```
Comment (Aratz) [inline arteria/models/runfolder_utils.py:129]: Where is this used and why is it needed?

File: arteria/models/runfolder_utils.py [hunk 13]
```diff
@@ -116,6 +125,17 @@ def metadata(self):
...
+
+    def to_dict(self):
+        return {
+            "metadata": self.metadata,
+            "path": self.path,
+            "service_version": __version__,
```
Comment (Aratz) [inline arteria/models/runfolder_utils.py:135]: `service_version` is only related to the view we give to the user, so it should be added to the response by the handlers (in `serialize_runfolder_path`)

**Review by Aratz — COMMENTED** [review]
There was a few things I overlooked in `runfolder_utils`, so I have a few extra comments for you to address

File: arteria/models/runfolder_utils.py [hunk 14]
```diff
@@ -22,26 +22,51 @@ def list_runfolders(monitored_directories, filter_key=lambda r: True):
...
-    for monitored_directory in monitored_directories:
-        monitored_dir_path = Path(monitored_directory)
-        for subdir in monitored_dir_path.iterdir():
-            try:
-                if filter_key(runfolder := Runfolder(monitored_dir_path / subdir)):
-                    runfolders.append(runfolder)
-            except AssertionError as e:
-                if e == f"File [Rr]unParameters.xml not found in runfolder {subdir}":
-                    continue
+    monitored_runfolders_paths = get_monitored_path_files(monitored_directories)
```
Comment (Aratz) [inline arteria/models/runfolder_utils.py:25]: The name `get_monitored_path_files` does not quite reflect what this function does. Maybe it could be `get_monitored_runfolders` instead?

File: arteria/models/runfolder_utils.py [hunk 15]
```diff
@@ -22,26 +22,51 @@ def list_runfolders(monitored_directories, filter_key=lambda r: True):
...
         self.config = Config(DEFAULT_CONFIG)
-        self.path = Path(path)
-        assert self.path.is_dir()
+
+        # If specific runfolder_name has been provided in request.
+        if runfolder_name:
+            runfolder_name = runfolder_name
+            path = Path(path).parent / runfolder_name
+        else:
+            runfolder_name = Path(path).name
```
[the remark this answers is not in the document]
Comment (nkongenelly) [reply arteria/models/runfolder_utils.py:65]: My thought here was that only one endpoint specifies the runfolder i.e(/runfolders/path/(/.*)

image

**Comment by nkongenelly** [comment]
> Thanks for addressing my comments :) Unfortunately I found some corner cases that would break the current version of the service. I've updated the tests accordingly.
> 
> There are also some comments from previous reviews that were not addressed.
> 
> If you want we could schedule a pair-programming sessions to fix this.

Hi @Aratz , thank you so much for the review.
Sorry for the previous comments I missed, I might have missed seeing them and need to look keenly (not intentional at all, hehe))

And I think it would be a good idea to pair program to implement the changes suggested

**Comment by nkongenelly** [comment]
Great pair programming session to solve the comments... Thanks

**Review by Aratz — APPROVED** [review]
All good!
